compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureGenerator.java

```diff
@@ -418,9 +420,53 @@ public void visit(BLangRecordTypeNode recordTypeNode) {
             rewrite(field, recordTypeNode.typeDefEnv);
         }
         recordTypeNode.restFieldType = rewrite(recordTypeNode.restFieldType, env);
+        generateClosuresForDefaultValuesInTypeRefs(recordTypeNode, typeSymbol);
         result = recordTypeNode;
     }
 
+    private void generateClosuresForDefaultValuesInTypeRefs(BLangRecordTypeNode recordTypeNode,
+                                                            BTypeSymbol typeSymbol) {
+        for (BLangType type : recordTypeNode.typeRefs) {
+            BRecordType recordType = (BRecordType) Types.getReferredType(type.getBType());
+            Map<String, BInvokableSymbol> defaultValues = ((BRecordTypeSymbol) recordType.tsymbol).defaultValues;
+            for (Map.Entry<String, BInvokableSymbol> defaultValue : defaultValues.entrySet()) {
+                String name = defaultValue.getKey();
+                if (((BRecordTypeSymbol) typeSymbol).defaultValues.containsKey(name)) {
+                    continue;
+                }
+                BInvokableSymbol symbol = defaultValue.getValue();
+                BLangInvocation invocation = getFunctionPointerInvocation(symbol);
+                String closureName = RECORD_DELIMITER + recordTypeNode.symbol.name.value + RECORD_DELIMITER + name;
+                generateClosureForDefaultValues(closureName, name, invocation, symbol.retType, typeSymbol);
+            }
+        }
+    }
+
+    private BLangInvocation getFunctionPointerInvocation(BInvokableSymbol symbol) {
+        BLangInvocation funcInvocation = (BLangInvocation) TreeBuilder.createInvocationNode();
+        funcInvocation.setBType(symbol.retType);
+        funcInvocation.symbol = symbol;
+        funcInvocation.name = ASTBuilderUtil.createIdentifier(symbol.pos, symbol.name.value);
+        funcInvocation.functionPointerInvocation = true;
+        return funcInvocation;
+    }
+
+    private void generateClosureForDefaultValues(String closureName, String paramName, BLangInvocation invocation,
```

Comment:
Can't we reuse existing logic? Only the default value expression changes, right?